# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [48]:
#load python packages

import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
from sklearn.cluster import KMeans
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [49]:
file='data/step3_output.csv'
df=pd.read_csv(file)
df.head()

Name       state  summit_elev  vertical_drop  \
0            Eaglecrest Ski Area      Alaska         2600           1540   
1               Hilltop Ski Area      Alaska         2090            294   
2            Sunrise Park Resort     Arizona        11100           1800   
3  Yosemite Ski & Snowboard Area  California         7800            600   
4               Donner Ski Ranch  California         8012            750   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       0  ...   
1      0        0.0          0          0     0       1  ...   
2      0        0.0          0          1     2       3  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          0     0       1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              640.0            60.0              45.0       44.0   
1               30.0            30.0             150.0       36.0   
2              800.0            80.0             115.0       49.0   
3               88.0             0.0             110.0       84.0   
4              505.0            60.0             163.0       82.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            350.0          47.0          53.0               90.0   
1             69.0          30.0          34.0              152.0   
2            250.0          74.0          78.0              104.0   
3            300.0          47.0          47.0              107.0   
4            400.0          75.0          75.0              170.0   

   NightSkiing_ac  clusters  
0             0.0         2  
1            30.0         0  
2            80.0         1  
3             0.0         1  
4             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [50]:
df = pd.concat([df, pd.get_dummies(df['state'])], axis=1).drop(['state'], axis =1)
print(df.shape)
df.head()

(149, 59)


Name  summit_elev  vertical_drop  trams  \
0            Eaglecrest Ski Area         2600           1540      0   
1               Hilltop Ski Area         2090            294      0   
2            Sunrise Park Resort        11100           1800      0   
3  Yosemite Ski & Snowboard Area         7800            600      0   
4               Donner Ski Ranch         8012            750      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       0       4  ...             0   
1        0.0          0          0     0       1       0  ...             0   
2        0.0          0          1     2       3       1  ...             0   
3        0.0          0          0     0       1       3  ...             0   
4        0.0          0          0     0       1       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 59 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [51]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [52]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y , test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [53]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [54]:
# Make a variable called y_pred 
#and assign it the result of calling predict() on our model variable with parameter X_test

y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [55]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [56]:
evs = explained_variance_score(y_test, y_pred)

print(evs)

mae = mean_absolute_error(y_test, y_pred)

print(mae)

-1.4074434229496098e+21
126613679510.69292


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [57]:
print(lm.intercept_)

32290737111.59002


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [58]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
summit_elev        6.043297e+00
vertical_drop      1.256096e+00
trams              1.587359e+13
fastEight          3.155598e+13
fastSixes          1.888129e+13
fastQuads          2.628732e+12
quad               3.299039e+12
triple             5.542886e+12
double             6.367827e+12
surface            6.453834e+12
total_chairs       1.144589e+13
Runs               1.464844e-01
TerrainParks       5.136719e-01
LongestRun_mi      1.404053e+00
SkiableTerrain_ac  7.324219e-01
Snow Making_ac     2.304688e-01
daysOpenLastYear   8.835449e-01
yearsOpen          8.525391e-01
averageSnowfall    4.912109e-01
AdultWeekday       1.042285e+01
projectedDaysOpen  8.447266e-02
NightSkiing_ac     1.104492e+00
clusters           2.250977e+00
Alaska             1.844670e+11
Arizona            1.308807e+11
California         3.818829e+11
Colorado           2.590949e+11
Connecticut        2.886763e+11
Idaho              2.590949e+11
Illinois           2.251552e+11
Indiana            1.349741e+08
Iowa               1.308807e+11
Maine              2.886763e+11
Maryland           1.308807e+11
Massachusetts      2.251552e+11
Michigan           4.362102e+11
Minnesota          3.613262e+11
Missouri           1.308807e+11
Montana            1.844670e+11
Nevada             1.844670e+11
New Hampshire      4.010997e+11
New Mexico         3.151291e+11
New York           4.677091e+11
North Carolina     2.590949e+11
Ohio               2.251552e+11
Oregon             2.251552e+11
Pennsylvania       4.191609e+11
Rhode Island       1.308807e+11
South Dakota       4.231343e-11
Tennessee          1.308807e+11
Utah               2.251552e+11
Vermont            3.151291e+11
Virginia           1.844670e+11
Washington         2.251552e+11
West Virginia      2.251552e+11
Wisconsin          4.010997e+11
Wyoming            2.251552e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [59]:
file="data/step3_output.csv"
df=pd.read_csv(file)


In [60]:
X=df.drop(['AdultWeekend','Name','state'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y=y.ravel()


In [61]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)


print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.3435502571993758
7.310668509672001


In [62]:
print(lm.intercept_)

56.46551583003612


In [63]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
summit_elev        2.444117e+00
vertical_drop      1.398537e+00
trams              1.776357e-15
fastEight          2.997602e-15
fastSixes          1.221245e-15
fastQuads          4.248972e-01
quad               1.181158e+00
triple             6.330072e-01
double             8.900527e-01
surface            3.240516e-01
total_chairs       8.618602e-01
Runs               1.138202e+00
TerrainParks       7.610112e-01
LongestRun_mi      2.655334e-01
SkiableTerrain_ac  1.237582e+00
Snow Making_ac     2.783784e-01
daysOpenLastYear   1.914972e+00
yearsOpen          2.335216e-01
averageSnowfall    1.632668e+00
AdultWeekday       1.137163e+01
projectedDaysOpen  1.644644e-01
NightSkiing_ac     1.149075e+00
clusters           3.619353e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [64]:
file="data/step3_output.csv"
df=pd.read_csv(file)

In [65]:
X=df.drop(['AdultWeekend','Name','state','summit_elev'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y=y.ravel()

In [66]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)


print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.32250904816245907
7.4175134883246185


In [67]:
print(lm.intercept_)

56.319771644192414


In [68]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
vertical_drop      1.200565e+00
trams              3.552714e-15
fastEight          4.440892e-16
fastSixes          4.440892e-16
fastQuads          1.574561e-01
quad               9.498019e-01
triple             7.018296e-01
double             1.002878e+00
surface            1.475630e-01
total_chairs       1.052211e+00
Runs               8.604457e-01
TerrainParks       6.927318e-01
LongestRun_mi      3.683346e-01
SkiableTerrain_ac  1.980981e+00
Snow Making_ac     8.946468e-02
daysOpenLastYear   1.649631e+00
yearsOpen          5.307178e-01
averageSnowfall    2.186870e+00
AdultWeekday       1.119302e+01
projectedDaysOpen  3.657520e-01
NightSkiing_ac     1.426835e+00
clusters           3.422886e+00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. |-1.4074434229496098e+21 | 126613679510.69292 |-|
| Model 2.     | 0.3435502571993758 |7.310668509672001 |'state'|
| Model 3.     |0.32250904816245907 |7.4175134883246185 |'state','summit_elev','base_elev'|

Model Selection:

According to the literature "Higher percentages of explained variance indicates a stronger strength of association. It also means that you make better predictions." Also lower MAE is better for the model therefore I would select the second model. 